### PR TITLE
Replace Tree control left click on title signal with new signal for either left or right click

### PR DESCRIPTION
**Pull request by derammo**
Right mouse button support for clicking on the "column title" headers in the Tree control was apparently missed when right click support was added for other items.

This interaction style is needed to support a context menu to configure hiding/showing of optional columns in a normal way.

@KoBeWi as discussed on chat.

### CODE REVIEW

**Comment by derammo** [comment]
@YuriSizov does 4.x mean deferred past 4.0 (i.e. not anytime soon?)  Because I need this for current development of the debugger.

**Comment by YuriSizov** [comment]
No, `4.x` means "any 4.x version". If this is reviewed and approved, it can be merged in 4.0. But as it stands, it doesn't seem essential.

The debugger work doesn't seem to be approved yet by any maintainer, so I don't know if we want it in 4.0 or if it would require breaking compatibility. But if https://github.com/godotengine/godot/pull/64364 is approved for 4.0, then this can be moved as well.

**Comment by derammo** [comment]
Thank you for the detailed explanation.   That sounds perfect.

**Comment by derammo** [comment]
And just for clarity: this adds a new signal and intentionally does not remove the previous version of the signal, so that it does not break compatibility.

**Comment by kleonc** [comment]
> And just for clarity: this adds a new signal and intentionally does not remove the previous version of the signal, so that it does not break compatibility. The legacy signal continues to be fired in the same cases as before.

If such new signal is gonna be added then it makes sense to do break compatibility and remove the old signal as it would be redundant. Such change of course would need to be introduced in 4.0 (because of breaking compatibility).

If you'd want to backport it to `3.x` then yes, the old signal in there would need to be not removed.

I suggest/prefer removing the old signal in this PR (just my opinion).

File: scene/gui/tree.cpp
Comment (kleonc) [inline scene/gui/tree.cpp:3328]: `pox.x` is `real_t` (either `float` or `double`) so the cast should be to `real_t`. But I don't think casting is needed at all, it would be impilcitly converted anyway.

Comment (derammo) [reply scene/gui/tree.cpp:3328]: Good point, I missed it was real_t.   The cast is there to clarify to the static code analyzer that this is intentional and not a mistaken narrowing.  Recent analyzer in MSVC tags this as a probable bug because of loss of precision in the implicit cast.

Will correct to the right typedef though!

**Comment by derammo** [comment]
> > And just for clarity: this adds a new signal and intentionally does not remove the previous version of the signal, so that it does not break compatibility. The legacy signal continues to be fired in the same cases as before.
> 
> If such new signal is gonna be added then it makes sense to do break compatibility and remove the old signal as it would be redundant. Such change of course would need to be introduced in 4.0 (because of breaking compatibility).
> 
> If you'd want to backport it to `3.x` then yes, the old signal in there would need to be not removed.
> 
> I suggest/prefer removing the old signal in this PR (just my opinion).

I don't think I can accept responsibility for hunting down all the code that uses the old signal and fixing it, especially at a time when things are supposed to stabilize for 4.0?   I don't know how to proceed.

**Comment by kleonc** [comment]
> I don't think I can accept responsibility for hunting down all the code that uses the old signal and fixing it, especially at a time when things are supposed to stabilize for 4.0?

Responsibility is also on the reviewers/approvers/merger, that's how I see it. And searching for the signal name in the whole codebase and going through all the findings doesn't look too hard / error prone. Especially since I've justed searched for that old signal (`column_title_pressed`) and seems like it's not used in the whole codebase (in occures only in the `Tree` itself and its docs). So removal seems trivial.

>I don't know how to proceed.

As I pointed out, it was just my opinion. Whether others will agree with me that it should go into 4.0 I don't know. So I guess you can just wait for other opinions / decision. If you want you could add a second commit removing the old signal to show what it's about (and later based on the decision you would either remove that commit or squash both commits into one).

**Comment by derammo** [comment]
> Responsibility is also on the reviewers/approvers/merger, that's how I see it. And searching for the signal name in the whole codebase and going through all the findings doesn't look too hard / error prone. Especially since I've justed searched for that old signal (`column_title_pressed`) and seems like it's not used in the whole codebase (in occures only in the `Tree` itself and its docs). So removal seems trivial.

But `Tree` is a public API class 

```GDREGISTER_CLASS(Tree)```

I thought that means people can be using it in their Games' UIs or Editor Plugins and we'd never know?

> As I pointed out, it was just my opinion. Whether others will agree with me that it should go into 4.0 I don't know. So I guess you can just wait for other opinions / decision. If you want you could add a second commit removing the old signal to show what it's about (and later based on the decision you would either remove that commit or squash both commits into one).

I don't have enough experience with the Godot dev process I guess.  I will just wait a while and see what other comments happen.  Maybe it is normal to just remove API or maybe I misunderstood what is public API.   I do appreciate your careful review!

File: scene/gui/tree.cpp
Comment (derammo) [reply scene/gui/tree.cpp]: Changed via force push

**Comment by kleonc** [comment]
> But `Tree` is a public API class 
> `GDREGISTER_CLASS(Tree)` 
> I thought that means people can be using it in their Games' UIs or Editor Plugins and we'd never know?

Yes and yes. But Godot 4.0 is not meant to be backwards compatibile with 3.x, there are already many compatibility breaking changes between these versions. And, since 4.0 is still in the alpha stage, new breaking changes can still be introduced. I think it's pointed out in each 4.0.alpha release note that the things can change (be removed/changed/added and so on). So it's not like the users are not aware of this. E.g. from [the latest release note](https://godotengine.org/article/dev-snapshot-godot-4-0-alpha-14):

>Be aware that during the alpha stage the engine is still not feature-complete or stable. There will likely be breaking changes between this release and the first [beta release](https://en.wikipedia.org/wiki/Software_release_life_cycle#Beta). Only the beta will mark the so-called "feature freeze".
>
>As such, we do not recommend porting existing projects to this and other upcoming alpha releases unless you are prepared to do it again to fix future incompatibilities. However, if you can port some existing projects and demos to the new version, that may provide a lot of useful information about critical issues still left to fix.

Anyway, feels like we're getting offtopic. I think I've already stated my point (I like the new signal and I think it should be a replacement for the old one, not an addition to it) and I have nothing more to say here for now.

**Comment by derammo** [comment]
Removed the previous signal as requested.  This PR will now need the `breaks_compat` label.

**Comment by kleonc** [comment]
> Removed the previous signal as requested. This PR will now need the `breaks_compat` label.

@YuriSizov Shouldn't the milestone be changed to `4.0` as well? I think it should be resolved before the beta.

**Comment by YuriSizov** [comment]
> Shouldn't the milestone be changed to `4.0` as well? I think it should be resolved before the beta.

Sure, we can switch the milestone back and forth 🙃 But it still needs a review more than anything else, then we can adjust the metadata.

File: scene/gui/tree.cpp
Comment (kleonc) [inline scene/gui/tree.cpp:3462]:
```suggestion
```
Don't think a commented out (and unfinished) code is needed.

**Review by kleonc — APPROVED** [review]
Code changes look fine. Signal rename and added additional parameter seem like an obvious improvement to me.

File: scene/gui/tree.cpp
Comment (kleonc) [reply scene/gui/tree.cpp:3462]: >or will that invalidate your approval if I force push?

Not sure. But even if it would invalidate my approval here I think that anyone would get that I've already approved the actual changes. :slightly_smiling_face:

Comment (akien-mga) [reply scene/gui/tree.cpp:3462]: Further updating a PR doesn't invalidate past approvals (at least not automatically - of course if the PR morphs into something completely new a human might ask for re-assessment, but here that's not the case).

Comment (derammo) [reply scene/gui/tree.cpp]: And so it was pushed.

File: scene/gui/tree.cpp
Comment (KoBeWi) [inline scene/gui/tree.cpp:3344]: I wonder if this is necessary 🤔
This could probably support all buttons, since you have to check what button was used anyway.

**Review by KoBeWi — APPROVED** [review]

File: scene/gui/tree.cpp
Comment (YuriSizov) [inline scene/gui/tree.cpp:3358]: Why a static cast? And furthermore, if there is a need for a static cast, why is there not one down below in a similar code doing similar stuff with the same types of data? But I don't think there is a need to begin with.

Comment (KoBeWi) [reply scene/gui/tree.cpp:3358]: It was discussed here: https://github.com/godotengine/godot/pull/64369#discussion_r945174667

File: scene/gui/tree.cpp
Comment (KoBeWi) [reply scene/gui/tree.cpp:3344]: No what I mean is that maybe we could just allow any button, as there is no reason to prevent it I think.

File: scene/gui/tree.cpp
Comment (YuriSizov) [reply scene/gui/tree.cpp:3358]: Well then the code below should also be changed.

File: scene/gui/tree.cpp
Comment (YuriSizov) [reply scene/gui/tree.cpp:3344]: Ah, fair enough.

Comment (derammo) [reply scene/gui/tree.cpp:3344]: @KoBeWi I wondered exactly the same thing.  I decided to leave in the restriction to be conservative.  No good scenarios why....   I thought maybe for the possibility of having some kind of consistent UI actions in the future that use a particular mouse button?

Comment (KoBeWi) [reply scene/gui/tree.cpp:3344]: But the signal emits either left or right button, so to use it you need to check what button you have received. Then it doesn't matter how many buttons it supports, as you already have filter out anything unnecessary.

File: scene/gui/tree.cpp
Comment (derammo) [reply scene/gui/tree.cpp:3358]: Oh, you mean because of the massive code duplication? :)     I will go add it there... :(

Comment (YuriSizov) [reply scene/gui/tree.cpp:3358]: Uh-huh :)

Comment (derammo) [reply scene/gui/tree.cpp:3358]: changed via force push

File: scene/gui/tree.cpp
Comment (derammo) [reply scene/gui/tree.cpp:3344]: I agree.  I just wondered if Godot itself in the future might want to implement what happens with other buttons for consistency?   I didn't say it was a weak reason :)

I do think that this can of worms is a proposal, because it would affect ALL the "clicked" events everywhere.

Comment (derammo) [reply scene/gui/tree.cpp:3344]: More support for your point:  we are already breaking compatibility, might as well make it nice and generic this time so we don't have to do that to people again after 4.0 is final.   So yeah, a proposal across all Control nodes to remove restrictions on which buttons are notified via signals?

Comment (YuriSizov) [reply scene/gui/tree.cpp:3344]: Let's just make a change here first. If there are other controls that report button index in signals which are restricted, we can evaluate it separately.

Comment (derammo) [reply scene/gui/tree.cpp:3344]: Even this class has more than one signal like that.  Those were all changed in a previous PR to use button index (restricted) except this one was missed.  I fixed it because I need it to select shown columns in the Thread debugger UI :).   Are you suggesting to widen THIS PR to change more than that?

(sorted by me just now for clarification, not what the code has)
```
	ADD_SIGNAL(MethodInfo("column_title_clicked", PropertyInfo(Variant::INT, "column"), PropertyInfo(Variant::INT, "mouse_button_index")));
	ADD_SIGNAL(MethodInfo("item_mouse_selected", PropertyInfo(Variant::VECTOR2, "position"), PropertyInfo(Variant::INT, "mouse_button_index")));
	ADD_SIGNAL(MethodInfo("custom_item_clicked", PropertyInfo(Variant::INT, "mouse_button_index")));
	ADD_SIGNAL(MethodInfo("item_double_clicked"));
	ADD_SIGNAL(MethodInfo("button_clicked", PropertyInfo(Variant::OBJECT, "item", PROPERTY_HINT_RESOURCE_TYPE, "TreeItem"), PropertyInfo(Variant::INT, "column"), PropertyInfo(Variant::INT, "id"), PropertyInfo(Variant::INT, "mouse_button_index")));
	ADD_SIGNAL(MethodInfo("empty_clicked", PropertyInfo(Variant::VECTOR2, "position"), PropertyInfo(Variant::INT, "mouse_button_index")));

	ADD_SIGNAL(MethodInfo("item_selected"));
	ADD_SIGNAL(MethodInfo("cell_selected"));
	ADD_SIGNAL(MethodInfo("multi_selected", PropertyInfo(Variant::OBJECT, "item", PROPERTY_HINT_RESOURCE_TYPE, "TreeItem"), PropertyInfo(Variant::INT, "column"), PropertyInfo(Variant::BOOL, "selected")));
	ADD_SIGNAL(MethodInfo("item_edited"));
	ADD_SIGNAL(MethodInfo("item_custom_button_pressed"));
	ADD_SIGNAL(MethodInfo("item_collapsed", PropertyInfo(Variant::OBJECT, "item", PROPERTY_HINT_RESOURCE_TYPE, "TreeItem")));
	ADD_SIGNAL(MethodInfo("check_propagated_to_item", PropertyInfo(Variant::OBJECT, "item", PROPERTY_HINT_RESOURCE_TYPE, "TreeItem"), PropertyInfo(Variant::INT, "column")));
	ADD_SIGNAL(MethodInfo("custom_popup_edited", PropertyInfo(Variant::BOOL, "arrow_clicked")));
	ADD_SIGNAL(MethodInfo("item_activated"));
	ADD_SIGNAL(MethodInfo("nothing_selected"));
```

Comment (YuriSizov) [reply scene/gui/tree.cpp:3344]: > Are you suggesting to widen THIS PR to change more than that?

No. This is already an extra comment on the approval from KoBeWi, so this is by no means a requirement.

**Comment by YuriSizov** [comment]
Thanks!

File: scene/gui/tree.cpp
Comment (derammo) [reply scene/gui/tree.cpp:3344]: Oh thank g*d*t. :)

**Pull request merged**
